### ISSUE TITLE
bad order of populations with read.structure() when pops are coded with number

### ISSUE THREAD

**Issue by melissalem** [thibautjombart/adegenet#142]
Hi,

I have a dataset converted in genepop and structure that differ by the fact that the pop names in the genepop file are factors while they are numbers in the structure file. If I donwload both files and calculate the pairwise Fst for each file, I get different results because the pops are re-organized when loading the data with read.structure.

 `obj<-read.structure(system.file("files/nancycats.str",package="adegenet"),onerowperind=FALSE, n.ind=237, n.loc=9, col.lab=1, col.pop=2, ask=FALSE)`
`table(pop(obj))`

 1 10 11 12 13 14 15 16 17  2  3  4  5  6  7  8  9 
10 11 20 14 13 17 11 12 13 22 12 23 15 11 14 10  9 

`obj <- read.genepop(system.file("files/nancycats.gen",package="adegenet"))`
`table(pop(obj))`

 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 
10 22 12 23 15 11 14 10  9 11 20 14 13 17 11 12 13 

How to code the pops in order to conserve the order?

Thanks a lot.

**Comment by jgx65** [comment]
Thanks for the report Melissa, I will look into it.  Note that the table(pop(obj)) seems to provide the correct info.  Which pairwise FST are you using? (`genet.dist` from `hierfstat`?)

**Comment by melissalem** [comment]
Dear Jerome

I was using pairwise_D from mmod.
But the problem is that when you use a integer for pop.names,
read.structure changes the order of the pops (1 then 10 and not 1 then 2).
I am now using a factor (A,B...) to code my pops and read.structure to
download my data in adegenet and my problem is solved (pairwise Fst
calculated with a genpop file and a structure are the same). But I still
think this is a bug in read.structure.

Perhaps only with the version I am using: Rv 3.2.3 adegent 2.0.1 or with
the combinations of packages that I am using??
I checked and I arrive to the same conclusions when I download only
adegenet.

Best,
Mél.


obj <- read.structure(system.file("files/nancycats.str",package="adegenet"),
  onerowperind=FALSE, n.ind=237, n.loc=9, col.lab=1, col.pop=2, ask=FALSE)

obj2 <- read.genepop(system.file("files/nancycats.gen",package="adegenet"))


> table(obj$pop)

 1 10 11 12 13 14 15 16 17  2  3  4  5  6  7  8  9
10 11 20 14 13 17 11 12 13 22 12 23 15 11 14 10  9
> table(pop(obj))

 1 10 11 12 13 14 15 16 17  2  3  4  5  6  7  8  9
10 11 20 14 13 17 11 12 13 22 12 23 15 11 14 10  9

> table(obj2$pop)

 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17
10 22 12 23 15 11 14 10  9 11 20 14 13 17 11 12 13
>
> table(pop(obj2))

 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17
10 22 12 23 15 11 14 10  9 11 20 14 13 17 11 12 13


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DDB4FAA8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

2016-05-09 8:46 GMT+02:00 Jerome Goudet <notifications@github.com>:

> Thanks for the report Melissa, I will look into it. Note that the
> table(pop(obj)) seems to provide the correct info. Which pairwise FST are
> you using? (genet.dist from hierfstat?)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/thibautjombart/adegenet/issues/142#issuecomment-217787831>
>



-- 
Mélissa Lemoine, PhD


Department of Evolutionary Biology and Environmental Studies
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich

**Comment by thibautjombart** [comment]
thanks for the report, looking into it today

**Comment by thibautjombart** [comment]
Fixed as of e43f675f36

@melissalem some explanations: the ordering of levels in 'pop' depends on the type of input, as you pointed out; this is because:
```r
> factor(1:11)
 [1] 1  2  3  4  5  6  7  8  9  10 11
Levels: 1 2 3 4 5 6 7 8 9 10 11
> factor(as.character(1:11))
 [1] 1  2  3  4  5  6  7  8  9  10 11
Levels: 1 10 11 2 3 4 5 6 7 8 9
```

I have reformed the code of import functions so that now, the conversion of `pop` to a `factor` is handled solely by the constructor, which is now forced to be consistent with the `pop` accessor. In a nutshell: levels of `pop` will now always be by the order of appearance of the populations in the input file. 

Consistency across different inputs have now been added to unit tests.

We now have:

```r
obj <- read.structure(system.file("files/nancycats.str",package="adegenet"),
  onerowperind=FALSE, n.ind=237, n.loc=9, col.lab=1, col.pop=2, ask=FALSE)

obj2 <- read.genepop(system.file("files/nancycats.gen",package="adegenet"))

table(pop(obj))

 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 
10 22 12 23 15 11 14 10  9 11 20 14 13 17 11 12 13 

table(pop(obj2))

 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 
10 22 12 23 15 11 14 10  9 11 20 14 13 17 11 12 13 
```

Cheers

**Comment by thibautjombart** [comment]
To check:
```r
devtools::install_github("thibautjombart/adegenet")
library(adegenet)
```
etc.
